internal/job/restore/restore_test.go

```diff
@@ -20,60 +19,92 @@ import (
+	defaultVolumeSize  = defaultMaxPartSize * 2
+	defaultChunkSize   = uint64(4096)
+)
+
 type setupInput struct {
+	// Split size when executing export-diff
+	maxPartSize uint64
 	// if true, run incremental backup after full backup, otherwise, only run full backup.
 	enableIncrementalBackup bool
+	// Virtual volume size when creating backups.
+	volumeSize []uint64
+	// Restore target block device size.
+	// If 0, use volumeSize.
+	blockDeviceSize []uint64
+	// chunk size for restore
+	chunkSize uint64
 	// if > 0, delete backups after creating them.
 	// Backups are deleted starting with the oldest ones, up to the specified number of backups.
 	deleteBackups int
 }
 
 type setupOutput struct {
+	nlvRepo model.NodeLocalVolumeRepository
 	// RestoreInput instance used when calling the restore module.
 	// 0 index is for the full backup, and 1 index is for the incremental backup.
 	restoreInputs []*input.Restore
+	volumeRaws    [][]byte
 }
 
 func setup(t *testing.T, config *setupInput) *setupOutput {
 	t.Helper()
 
-	k8sRepo, rbdRepo, volumeInfo := fake.NewStorage()
+	k8sRepo, _, volumeInfo := fake.NewStorage()
+	rbdRepo := fake.NewRBDRepository2(volumeInfo.PoolName, volumeInfo.ImageName)
 	nlvRepo, finRepo, _ := testutil.CreateNLVAndFinRepoForTest(t)
 
-	rawImageChunkSize := uint64(4096)
-	targetSnapshotSize := rawImageChunkSize * 2
-
 	snapIDs := make([]int, 0)
 	restoreInputs := make([]*input.Restore, 0)
+	volumeRaws := make([][]byte, 0)
 
+	if config.maxPartSize == 0 {
+		config.maxPartSize = defaultMaxPartSize
+	}
+	if config.chunkSize == 0 {
+		config.chunkSize = defaultChunkSize
+	}
 	// Create backups.
 	backupCount := 1
 	if config.enableIncrementalBackup {
 		backupCount = 2
 	}
 	for i := 0; i < backupCount; i++ {
-		snapshot := rbdRepo.CreateFakeSnapshot(utils.GetUniqueName("snap-"), targetSnapshotSize, time.Now())
+		volumeSize := defaultVolumeSize
+		if len(config.volumeSize) > i {
+			volumeSize = config.volumeSize[i]
+		}
+		snapshot, volumeRaw, err := rbdRepo.CreateSnapshotWithRandomData(utils.GetUniqueName("snap-"), volumeSize)
+		require.NoError(t, err)
 		snapIDs = append(snapIDs, snapshot.ID)
+		volumeRaws = append(volumeRaws, volumeRaw)
 		var srcSnapID *int
 		if i != 0 {
 			srcSnapID = &snapIDs[i-1]
 		}
-		backupInput := testutil.NewBackupInput(k8sRepo, volumeInfo, snapshot.ID, srcSnapID, rawImageChunkSize)
+		backupInput := testutil.NewBackupInput(k8sRepo, volumeInfo, snapshot.ID, srcSnapID, config.maxPartSize)
 		backupInput.Repo = finRepo
 		backupInput.KubernetesRepo = k8sRepo
 		backupInput.RBDRepo = rbdRepo
 		backupInput.NodeLocalVolumeRepo = nlvRepo
 
 		bk := backup.NewBackup(backupInput)
-		err := bk.Perform()
+		err = bk.Perform()
 		require.NoError(t, err)
 
 		// Create the restore file
-		restorePath := testutil.CreateRestoreFileForTest(t, targetSnapshotSize)
-		rVol := fake.NewRestoreVolume(restorePath)
+		blockDeviceSize := volumeSize
+		if len(config.blockDeviceSize) > i && config.blockDeviceSize[i] != 0 {
```

Comment:
Why do we need `config.blockDeviceSize[i] != 0` condition?

Reply:
It's not in use, and I removed it.